### PR TITLE
feat: [Table] dfk_hero_sales

### LINKED ISSUE #21: Table: DFK_Hero_Transactions
Let's create a `DFK_Hero_Transactions` table to trace all DFK Hero Purchases.
This can be created using `txs+logs` as the source data table.

Hints:
```
with buy_hero_tx AS (
select block_timestamp, tx_id, event_inputs:tokenId as token_id from harmony.dev.ant_logs
where 1=1
and event_inputs:tokenId is not null
and event_emitted_name != 'Approval'
and event_inputs:from = '0x13a65b9f8039e2c032bc022171dc05b30c3f2892'
and event_inputs:to = :1
)
select b.block_timestamp, n.tx_id, jewels_paid, token_id, nvl(p.price*jewels_paid,0) as amount_usd from (
select 
block_timestamp,
tx_id,
sum(event_inputs:value / pow(10,18)) as jewels_paid
from harmony.dev.ant_logs
where tx_id in (select tx_id from buy_hero_tx)
and event_emitted_name != 'Approval'
group by tx_id, block_timestamp
) n
LEFT JOIN buy_hero_tx b ON b.tx_id = n.tx_id
LEFT JOIN harmony.dev.ant_view_token_price_jewel p ON date(b.block_timestamp) = date(timestamp)
where jewels_paid is not null
order by b.block_timestamp asc
```

**To Do Next:**

- [ ] Design the schema for this table
- [ ] Get confirmation on schema
- [ ] Build schema
- [ ] Schema build review
- [ ] Submit Pull Request

### PR DESCRIPTION
**Pull request by antonyip**
Closes #21

### CODE REVIEW

File: models/dfk/dfk_hero_sales.sql [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        block_timestamp,
+        java_hextoint(replace(topics[1]::string, '0x','' )) as hero_token_id,
+        java_hextoint(substr(data, 3, 64)) as auction_id,
+        java_hextoint(substr(data, 3+64, 64))::number / 1e18  as total_jewels,
+        java_hextoint(substr(data, 3+64, 64))::number / 1e18 * 0.0375 as tax_jewels,
+        concat('0x',LTRIM(substr(data, 3+128, 64),'0')) as buyer_address,
+        total_jewels * jewel_price.price  as total_usd,
+        tax_jewels * jewel_price.price as tax_usd
+    from logs
+    left join harmony.dev.tokenprice_jewel as jewel_price
```
Comment (chuxinh) [inline models/dfk/dfk_hero_sales.sql]: update to prod ref

File: models/dfk/dfk_hero_sales.sql [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        on date_trunc('day', block_timestamp) = jewel_price.timestamp
+    where event_removed = false
```
Comment (chuxinh) [inline models/dfk/dfk_hero_sales.sql]:
```suggestion
    where not event_removed
```

File: models/dfk/dfk_hero_sales.sql [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        tx_hash,
+        block_timestamp,
+        java_hextoint(replace(topics[1]::string, '0x','' )) as hero_token_id,
+        java_hextoint(substr(data, 3, 64)) as auction_id,
+        java_hextoint(substr(data, 3+64, 64))::number / 1e18  as total_jewels,
+        java_hextoint(substr(data, 3+64, 64))::number / 1e18 * 0.0375 as tax_jewels,
+        concat('0x',LTRIM(substr(data, 3+128, 64),'0')) as buyer_address,
+        total_jewels * jewel_price.price  as total_usd,
+        tax_jewels * jewel_price.price as tax_usd
+    from logs
```
Comment (chuxinh) [inline models/dfk/dfk_hero_sales.sql:43]: recommend add alias to this and the columns above

File: models/dfk/dfk_hero_sales.sql [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    left join harmony.dev.tokenprice_jewel as jewel_price
+        on date_trunc('day', block_timestamp) = jewel_price.timestamp
+    where event_removed = false
+        and evm_contract_address = '0x13a65b9f8039e2c032bc022171dc05b30c3f2892'
```
Comment (chuxinh) [inline models/dfk/dfk_hero_sales.sql]: add comment on this address

File: models/dfk/dfk_hero_sales.sql [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        and topics[0]::string ='0xe40da2ed231723b222a7ba7da994c3afc3f83a51da76262083e38841e2da0982' -- AuctionSuccessful event
+),
+final as (
+    select
+        auction_created.seller_address,
+        hero_auction_txns.*
+    from auction_created
+    right join hero_auction_txns
```
Comment (chuxinh) [inline models/dfk/dfk_hero_sales.sql]: left join?

**Review by chuxinh — CHANGES_REQUESTED** [review]

File: models/dfk/dfk_hero_sales.sql (hunk 3 above)
Comment (antonyip) [reply models/dfk/dfk_hero_sales.sql:43]: Not sure what this meant...

File: models/dfk/dfk_hero_sales.sql [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    where not event_removed
+        and evm_contract_address = '0x13a65b9f8039e2c032bc022171dc05b30c3f2892' -- Serendale_AuctionHouse
+        and topics[0]::string ='0xe40da2ed231723b222a7ba7da994c3afc3f83a51da76262083e38841e2da0982' -- AuctionSuccessful event
+),
+final as (
+    select
+        hero_auction_txns.*,
+        auction_created.seller_address
+    from hero_auction_txns
+    left join auction_created
```
Comment (chuxinh) [inline models/dfk/dfk_hero_sales.sql:55]:
```suggestion
    join auction_created
```

File: models/dfk/dfk_hero_sales.sql (hunk 3 above)
Comment (chuxinh) [reply models/dfk/dfk_hero_sales.sql:43]: oh such as logs.tx_hash or if you put alias `log as l` then `l.tx_hash`
as it's not clear which table the column is coming from when you join

**Review by chuxinh — APPROVED** [review]
LGTM after nits

File: models/dfk/dfk_hero_sales.sql (hunk 3 above)
Comment (antonyip) [reply models/dfk/dfk_hero_sales.sql:43]: hmm, but this is the alias from its own table, so I'm not too sure how to alias this?
(Will have the same problem as https://github.com/MetricsDAO/harmony_dbt/pull/82/files)

**Pull request merged**